stdlib/LinearAlgebra/src/matmul.jl

```diff
@@ -260,7 +264,13 @@ julia> C
  730.0  740.0
 ```
 """
-@inline mul!(C::AbstractMatrix, A::AbstractVecOrMat, B::AbstractVecOrMat, α::Number, β::Number) =
+@inline function mul!(C::AbstractMatrix, A::AbstractVecOrMat, B::AbstractVecOrMat, α::Number, β::Number)
+    if (!((eltype(C) <: BlasFloat && eltype(C) === eltype(A) === eltype(B)))) ||
+        (!(C isa StridedVecOrMat && A isa StridedVecOrMat && B isa StridedVecOrMat)) &&
+         α == true && (β == true || β == false)
```

Comment:
IIUC, then this condition applies, for instance, to CUDA, CUSparse and sparse matrices (and/or their transposes and adjoints), right? That would massively slow down SparseArrays.jl and even break a few GPUArrays-related packages, which specifically overload `LinearAlgebra.generic_matmatmul!` (with the character arguments).

Reply:
What are their overloads for non-`BlasFloat`?

Reply:
I can make it `Array`-only.

Reply:
Wait, I might have misread the logic. Suppose I want to multiply non-transposed dense GPUArrays with `BlasFloat` elements, then this returns ... true? And directs away from `generic_matmatmul!`? Too many conditions for my little brain.

Reply:
The check is supposed to be for BLAS compatibility of types. 
BLAS requires both of
1.The eltypes match and they're BlasFloat
2. They're all strided

But extending in this way to avoid dispatches assumes people aren't overloading these internal methods. I don't think Pele should, but apparently people do (I did confirm that CUDA does).

Reply:
https://github.com/JuliaGPU/CUDA.jl/blob/301318e44b0d99e63bc21c82d9038ff0dfdb21f4/lib/cublas/linalg.jl#L286

Reply:
Yes, they do because that significantly reduced load times of those packages, both SparseArrays and all of the GPUArrays-related packages. Otherwise people had to overload tons of `mul!` methods, for all combinations of plain/transpose/adjoint factors, possibly dispatching on type parameters. https://github.com/JuliaGPU/CUDA.jl/pull/1904

Reply:
Okay, ideally there would have been another way to avoid ``mul!`. 

I went with requiring `Matrix` for the first two arguments and `Array` for the third.

Reply:
> Okay, ideally there would have been another way to avoid `mul!`.

Which one? Since this is all internal, it can be changed to whatever yields the correct dispatch to the underlying *BLAS methods and doesn't increase load times. Method insertion at package loading was the big big issue.